docs/design/0012-generic-association.md

```diff
@@ -0,0 +1,275 @@
+
+## Detailed Problem Statement
+
+### Specify how an application is associated with an Elastic Stack Application
+
+The first problem is to make the association information available to the associated application. Each associated application may require a different set of information. For instance some of them may want to establish a connection to Elasticsearch while others may require the APM token and the APM Server certs to upload some metrics.
+
+Two kind of artifacts can be used to describe an association _(not exclusive)_:
+  1. A **new CRD** which would allow to specify the association between the associated application and some Elastic stack applications 
+  2. One or several ***annotations***, set on top-level objects resources that manage the associated application Pods. 
+
+### Inject the required artifacts into a Pod
+
+For the record there a 3 ways to inject some data into an application:
+
+  - [Environment variables](https://kubernetes.io/docs/tasks/inject-data-application/define-environment-variable-container/), examples of good candidates to be injected with environment variables are:
+    - Elastic stack application urls
+    - Username and Password
+    - APM Server token
+
+  - [Secrets](https://kubernetes.io/docs/tasks/inject-data-application/distribute-credentials-secure/) _(not mutually exclusive with environment variables)_, they naturally address the case of the certificates, but they can also contain the variables exposed as environment variables.
+ 
+  - [PodPreset](https://kubernetes.io/docs/tasks/inject-data-application/podpreset/), it is an interesting solution, designed to solve the problem described here. Unfortunately this is a recent addition in the Kubernetes API and it's not widely available yet.
+
+## Considered Options
+
+Since [PodPreset](https://kubernetes.io/docs/tasks/inject-data-application/podpreset/) is not available yet we are only considering here a mixed usage of a new CRD to specify the association information to be exposed plus some annotations on the top-level resources to let user specify how ECK should take over some part of the `Pod` spec.
+
+### Combine a new CRD with an annotation
+
+In this option a new CRD is introduced. It allows an Elastic stack application owner to specify for some third party applications which association data is required.
+
+#### Association specification as a new CRD
+
+```yaml
+apiVersion: association.k8s.elastic.co/v1alpha1
+kind: Association
+metadata:
+  name: monitor
+  namespace: monitor-ns # the Elastic stack applicaitons must live in this namespace
+spec:
+
+  ## targetNamespaces are the namespaces where the association information are copied and reconciled
+  targetNamespaces:
+    - ns1
+
+  ## These are the Elastic applications.
+  ## The namespace is automatically inherited from this custom resource.
+  elasticsearchName: monitor
+  kibanaName: monitor
+  # apmServerName: monitor
+  
```

Comment:
The rational here was to allow a quick review of the existing associations (for example in case of a security audit) and to be allow to have an alpha release and improve this feature over the time (decoupled lifecycle with the stack applications). But this could be discussed.